### PR TITLE
optimize txpool

### PR DESCRIPTION
**Pull request by yann-sjtu**

### CODE REVIEW

File: zk/stages/stage_sequence_execute_utils.go [hunk 1]
```diff
@@ -44,7 +44,7 @@ const (
 
 	transactionGasLimit = 30000000
 
-	yieldSize = 100 // arbitrary number defining how many transactions to yield from the pool at once
+	yieldSize = 1000 // arbitrary number defining how many transactions to yield from the pool at once
```
Comment (louisliu2048) [inline zk/stages/stage_sequence_execute_utils.go:47]: our block can only consume about 350 tx/block now, maybe this param should be set to 500 and later use apollo to support dynamic change.

File: zk/txpool/pool.go [hunk 2]
```diff
@@ -1233,8 +1258,8 @@ func promote(pending *PendingPool, baseFee, queued *SubPool, pendingBaseFee uint
 		}
 	}
 
-	// Promote best transactions from base fee pool to pending pool while they qualify
-	for best := baseFee.Best(); baseFee.Len() > 0 && best.subPool >= BaseFeePoolBits && best.minFeeCap.Cmp(uint256.NewInt(pendingBaseFee)) >= 0; best = baseFee.Best() {
+	// Promote best transactions from base fee pool to pending pool while they qualify and pending pool is not full
+	for best := baseFee.Best(); baseFee.Len() > 0 && !pending.IsFull() && best.subPool >= BaseFeePoolBits && best.minFeeCap.Cmp(uint256.NewInt(pendingBaseFee)) >= 0; best = baseFee.Best() {
 		tx := baseFee.PopBest()
 		announcements.Append(tx.Tx.Type, tx.Tx.Size, tx.Tx.IDHash[:])
 		pending.Add(tx)
```
Comment (louisliu2048) [inline zk/txpool/pool.go:1265]: maybe we should also add IsFull() judgment to queued pool when do promote

File: zk/txpool/pool.go [hunk 3]
```diff
@@ -2082,6 +2101,9 @@ func (p *PendingPool) Worst() *metaTx { //nolint
 }
 func (p *PendingPool) PopWorst() *metaTx { //nolint
 	i := heap.Pop(p.worst).(*metaTx)
+	if i.bestIndex != p.Len()-1 { // which should never happen
```
Comment (louisliu2048) [inline zk/txpool/pool.go:2104]: when we popWorst from pendingPool, I think the pending pool still be sorted.

File: zk/txpool/pool_zk.go [hunk 4]
```diff
@@ -158,6 +158,8 @@ func (p *TxPool) best(n uint16, txs *types.TxsRlp, tx kv.Tx, onTopOf, availableG
 	isShanghai := p.isShanghai()
 	isLondon := p.isLondon()
 	_ = isLondon
+
+	p.pending.EnforceBestInvariants() // it costs about 50ms when pending size reached one million
```
Comment (louisliu2048) [inline zk/txpool/pool_zk.go:162]: This interface will be called by grpc server, maybe we should also forbiden that grpc api

Comment (yann-sjtu) [reply zk/txpool/pool_zk.go:162]: will check that

**Review by ylsGit — APPROVED** [review]

**Review by chengzhinei — APPROVED** [review]

**Review by louisliu2048 — APPROVED** [review]

**Review by chengzhinei — APPROVED** [review]

**Pull request merged**